autowarefoundation/autoware
# Modify a way to calculate cost in costmap generator

## New feature implementation

### Implemented feature

Feature request issue. autowarefoundation/autoware_ai#641 



### Implementation description


In `costmap_generator`, first expand existing object. Second, applying smooth-out kernel to get the costmap.
![image](https://user-images.githubusercontent.com/10920881/57509075-47b93f00-733e-11e9-8b7f-7de6640d60dc.png)

The current implementation in `costmap_generator` smooths out the cost in an expanded area. 

However, this implementation is not good in terms of Planning.
In this PR, `costmap_generator` keep the cost highest in the expanded grid area.

#### How to test the new feature
1. Download the rosbag which includes dummy objects in topic:`/detection/lidar_detector/objects` from [here](https://drive.google.com/file/d/1uGT-ZdSR5BEeP85eXc9iAH5OyvqXolrl/view?usp=sharing).

2. Open run time manager.

3. Click `app` next to `costmap_generator`.

4. Check `use_objects_box`. Change `objects_input` to `/detection/lidar_detector/objects`. Change `expand_polygon_size` to `1`. Change `size_of_expansion_kernel` to `9`

![image](https://user-images.githubusercontent.com/10920881/57509635-b6e36300-733f-11e9-995d-9ed9a6ec145f.png)


5. Close `app` and launch `costmap_generator`. 

6. In rviz, checking subscription to `/semantics/costmap_generator/occupancy_grid` in Occupancy Grid Map.
![image](https://user-images.githubusercontent.com/10920881/57509935-759f8300-7340-11e9-9fc9-ffedd9310574.png)

7. You can see a result like the below image.
![modified_3](https://user-images.githubusercontent.com/10920881/57509991-9536ab80-7340-11e9-97a4-f9bc8a23795c.png)
